### ISSUE TITLE
IBM Websphere MQ Queue reads remove messages from the queue

### ISSUE THREAD

**Issue by jay-tb** [jjjw420/splunk-ibm-websphere-mq-add-on#7]
When connecting the _IBM Websphere MQ Queue_ input, during our testing messages are removed from the queue.  Is this intended?  

According to [this article on StackExchange](https://stackoverflow.com/questions/1039240/how-do-i-browse-a-websphere-mq-message-without-removing-it), it seems the ** MQOO_BROWSE** option makes it possible to monitor without removal.  

Might you share your use-case where Splunk is your processor for the messages it reads?

**Comment by jjjw420** [comment]
Indeed.   Working as intended.   I will gladly add a browse option to the input though.  Trivial to do and not really the problem.   

Consider...  Browsing the queue has implications that may not be suitable for every use case.  
Because the messages stays on the queue after being read there is nothing stops the splunk input from coming round on it's next run interval (like 60 seconds later) and reading the same message again causing duplicate events in Splunk - especially if indexing using the MQMD PUT time - which is the default.   You can tell the response handler to index using current time - but you will still be indexing the same message potentially more than once.    

I have thought about keeping track of the last browsed messageid and index from that message onward the next time the input runs - but that requires state to be saved for every queue (which seems inevitable).    Even a "persistent" handle to the queue will not be enough as the input will still need to know where it left off in the event that it's restarted or the connection to MQ is dropped.   

This is why I have not added a "browse" option to date.  I will see what the options are around this. 
If you have ideas of how you think this should work - please let me know.

**Comment by jjjw420** [comment]
After writing the above reply and giving it some thought.   Adding a browse option is comes with challenges and is possible - but will unfortunately require state to be saved if the intention is not to index the same message more than once when it's browsed multiple times by the input.

So....
- I will add a "browse" option to the queue input.   
- I will add a "save state after browse" option (which will be selected by default) to select if the input should continue from where it left off on the last run (eg.  only start indexing new messages not previosly indexed) or simply index all messages on the queue every time the input runs.

If state has to be saved:
- After every run per queue I will save the MQMD.MessageId of the last browsed message(eg. the last message on the queue).  
- On the next run I will browse the queue until I find the message with the saved MQMD.MessageId and start indexing the messages that follow after that.   If the MessageId is not found then all the messages in the queue must be indexed.
- I will add some text so that the implication of this browse action is fully understood.

If state is not to be saved the input will simply index the messages on the queue from start to finish every time.

Browsing the queue looking for the last message that was indexed requires processing time(eg CPU).      So this action will require some footwork to ensure that the queue is browsed as few times as possible(eg. preferably only once) - especially if the queue contains all new messages.   I can see this solution working well and without issues with queues containing hundreds of messages or thousands of messages.   But when the number of messages in the queue is increased to 10000 or more then much additional processing will be required as the queue is browsed from the beginning every time in order to find new messages.  I will see how feasable a persistent handle to the queue really is too.

I hope all of this makes sense.   If it does not - let me know.

I will get something out soonest.
